Fails to Detect Installed Versions

This is a snippet from a GitHub Action where I manually install the `tailwindcss-cli` and invoke the help function to show version, and then immediately call `trunk tools show`. Notice that it doesn't recognize the installed tailwind version, even though it would be the executable version on the path.

```sh
Run curl -sLO https://github.com/tailwindlabs/tailwindcss/releases/latest/download/tailwindcss-linux-x64
  curl -sLO https://github.com/tailwindlabs/tailwindcss/releases/latest/download/tailwindcss-linux-x64
  sudo mv tailwindcss-linux-x64 /usr/local/bin/tailwindcss
  chmod +x /usr/local/bin/tailwindcss
  tailwindcss --help
  trunk tools show
  shell: /usr/bin/bash -e {0}
  env:
    CARGO_TERM_COLOR: always
    SQLX_OFFLINE: true
  
≈ tailwindcss v4.0.15
Usage:
  tailwindcss [--input input.css] [--output output.css] [--watch] [options…]
Options:
  -i, --input ··········· Input file
  -o, --output ·········· Output file [default: `-`]
  -w, --watch ··········· Watch for changes and rebuild as needed
  -m, --minify ·········· Optimize and minify the output
      --optimize ········ Optimize the output without minifying
      --cwd ············· The current working directory [default: `.`]
  -h, --help ············ Display usage information
sass
    Installed Version: n/a
    Default Version: 1.69.5
    Download URL: https://github.com/sass/dart-sass/releases/download/1.69.5/dart-sass-1.69.5-linux-x64.tar.gz
    Location: n/a
tailwindcss
    Installed Version: n/a
    Default Version: 3.3.5
    Download URL: https://github.com/tailwindlabs/tailwindcss/releases/download/v3.3.5/tailwindcss-linux-x64
    Location: n/a
tailwindcss-extra
    Installed Version: n/a
    Default Version: 1.7.25
    Download URL: https://github.com/dobicinaitis/tailwind-cli-extra/releases/download/v1.7.25/tailwindcss-extra-linux-x64
    Location: n/a
wasm-bindgen
    Installed Version: 0.2.100
    Default Version: 0.2.89
    Download URL: https://github.com/rustwasm/wasm-bindgen/releases/download/0.2.89/wasm-bindgen-0.2.89-x86_64-unknown-linux-musl.tar.gz
    Location: /home/runner/.cargo/bin/wasm-bindgen
wasm-opt
    Installed Version: n/a
    Default Version: version_116
    Download URL: https://github.com/WebAssembly/binaryen/releases/download/version_116/binaryen-version_116-x86_64-linux.tar.gz
    Location: n/a
```

Later, when the workflow tries to build the `frontend`, it fails to download tools.

```sh
Run trunk build --release
  trunk build --release
  shell: /usr/bin/bash -e {0}
  env:
    CARGO_TERM_COLOR: always
    SQLX_OFFLINE: true
    CACHE_ON_FAILURE: false
    CARGO_INCREMENTAL: 0
2025-03-23T17:06:28.777172Z  INFO 🚀 Starting trunk 0.21.9
2025-03-23T17:06:28.777559Z  INFO 📦 starting build
2025-03-23T17:06:30.187284Z  INFO downloading tailwindcss-extra version="2.0.7"
   Compiling proc-macro2 v1.0.94
   Compiling unicode-ident v1.0.18
   Compiling wasm-bindgen-shared v0.2.100
   Compiling bumpalo v3.17.0
   Compiling log v0.4.26
   Compiling rustversion v1.0.20
   Compiling once_cell v1.21.1
   Compiling wasm-bindgen v0.2.100
2025-03-23T17:06:30.922552Z  INFO installing tailwindcss-extra
   Compiling cfg-if v1.0.0
   Compiling serde v1.0.219
   Compiling memchr v2.7.4
   Compiling quote v1.0.40
   Compiling syn v2.0.100
   Compiling itoa v1.0.15
   Compiling ryu v1.0.20
   Compiling serde_json v1.0.140
   Compiling thiserror v1.0.69
   Compiling futures-sink v0.3.31
   Compiling futures-core v0.3.31
   Compiling autocfg v1.4.0
   Compiling futures-channel v0.3.31
   Compiling pin-project-lite v0.2.16
   Compiling pin-utils v0.1.0
   Compiling equivalent v1.0.2
   Compiling futures-io v0.3.31
   Compiling bytes v1.10.1
   Compiling slab v0.4.9
   Compiling hashbrown v0.15.2
   Compiling percent-encoding v2.3.1
   Compiling futures-task v0.3.31
   Compiling version_check v0.9.5
   Compiling fnv v1.0.7
   Compiling http v0.2.12
   Compiling indexmap v2.8.0
   Compiling form_urlencoded v1.2.1
   Compiling toml_datetime v0.6.8
   Compiling winnow v0.5.40
   Compiling gloo-utils v0.1.7
   Compiling gloo-events v0.1.2
   Compiling gloo-events v0.2.0
   Compiling gloo-console v0.2.3
   Compiling gloo-history v0.1.5
   Compiling gloo-history v0.2.2
   Compiling gloo-worker v0.2.1
   Compiling gloo-file v0.3.0
   Compiling gloo-file v0.2.3
   Compiling gloo-net v0.3.1
   Compiling gloo-storage v0.2.2
   Compiling gloo-storage v0.3.0
   Compiling gloo-console v0.3.0
   Compiling gloo-dialogs v0.1.1
   Compiling gloo-render v0.2.0
   Compiling gloo-render v0.1.1
   Compiling gloo-dialogs v0.2.0
   Compiling gloo v0.8.1
   Compiling gloo-net v0.4.0
   Compiling gloo-worker v0.4.0
   Compiling gloo v0.10.0
   Compiling prokio v0.1.0
   Compiling gloo-net v0.5.0
   Compiling yew v0.21.0
   Compiling gloo-worker v0.5.0
   Compiling gloo v0.11.0
   Compiling yewdux v0.11.0
   Compiling yew-hooks v0.3.3
   Compiling i18nrs v0.1.3
   Compiling yew-router v0.18.0
   Compiling yew_icons v0.8.0
   Compiling shared v0.1.0 (/home/runner/work/RustyGPT/RustyGPT/shared)
   Compiling frontend v0.1.0 (/home/runner/work/RustyGPT/RustyGPT/frontend)
    Finished `release` profile [optimized] target(s) in 54.14s
2025-03-23T17:07:24.924867Z ERROR ❌ error
error from build pipeline
Caused by:
    0: HTML build pipeline failed (1 errors), showing first
    1: error from asset pipeline
    2: Could not extract files
    3: error while getting archive entry
    4: invalid gzip header
2025-03-23T17:07:24.924893Z ERROR error from build pipeline
2025-03-23T17:07:24.924904Z  INFO   1: HTML build pipeline failed (1 errors), showing first
2025-03-23T17:07:24.924906Z  INFO   2: error from asset pipeline
2025-03-23T17:07:24.924908Z  INFO   3: Could not extract files
2025-03-23T17:07:24.924909Z  INFO   4: error while getting archive entry
2025-03-23T17:07:24.924911Z  INFO   5: invalid gzip header
Error: Process completed with exit code 1.
```

I've followed this issue down to https://github.com/trunk-rs/trunk/blob/e30000529606f2f5fb62c46a9f8413d7678dcf54/src/tools.rs#L368 where it calls into `which` and followed that down to [`which::which` -> `which_all`](https://github.com/harryfei/which-rs/blob/bd868818bdb55923acfd2a63468335f6600a6cf9/src/lib.rs#L92) where it's strictly bound to the `PATH` variable. So, I modified my workflows to ensure the directory containing the executables in question would be discovered:

```yaml
      - name: Add Cargo bin to PATH
        run: echo "/usr/local/cargo/bin" >> $GITHUB_PATH

      - name: Build frontend
        working-directory: frontend
        env:
          PATH: /usr/local/cargo/bin:${{ env.PATH }}
        run: |
          trunk build --release
```

According to my install script:

```sh
#!/bin/bash
set -e

# Sort out OS
OS=$(uname -s)
if [[ "$OS" == "Linux" ]]; then
  OS_TYPE="linux"
elif [[ "$OS" == "Darwin" ]]; then
  OS_TYPE="macos"
elif [[ "$OS" == *"MINGW"* || "$OS" == *"CYGWIN"* ]]; then
  OS_TYPE="windows"
else
  echo "Unsupported OS: ${OS}"
  exit 1
fi
echo "OS: ${OS_TYPE}"

# Sort out architecture
ARCH=$(uname -m)
echo "Architecture: ${ARCH}"

if [[ "$ARCH" == "x86_64" ]]; then
  ARCH="x64"
elif [[ "$ARCH" == "aarch64" || "$ARCH" == "arm64" ]]; then
  ARCH="arm64"
else
  echo "Unsupported architecture: ${ARCH}"
  exit 1
fi

# Set installation path
INSTALL_PATH="/usr/local/cargo/bin"

# Install base tailwindcss
BASE_ASSET="tailwindcss-${OS_TYPE}-${ARCH}"

echo "Fetching latest tailwind-cli release version..."
BASE_RELEASE_VERSION=$(curl -s https://api.github.com/repos/tailwindlabs/tailwindcss/releases/latest | jq -r '.tag_name')
echo "Latest tailwind-cli version: ${BASE_RELEASE_VERSION}"

echo "Installing tailwind-cli: ${BASE_RELEASE_VERSION} - ${BASE_ASSET}"
BASE_URL="https://github.com/tailwindlabs/tailwindcss/releases/download/${BASE_RELEASE_VERSION}/${BASE_ASSET}"
echo "Downloading tailwindcss from ${BASE_URL}"
curl -sLO "${BASE_URL}"
mv "${BASE_ASSET}" "$INSTALL_PATH/tailwindcss"
chmod +x "$INSTALL_PATH/tailwindcss"
$INSTALL_PATH/tailwindcss --help

# Install tailwindcss-extra
EXTRA_ASSET="tailwindcss-extra-${OS_TYPE}-${ARCH}"

echo "Fetching latest tailwind-cli-extra release version..."
EXTRA_RELEASE_VERSION=$(curl -s https://api.github.com/repos/dobicinaitis/tailwind-cli-extra/releases/latest | jq -r '.tag_name')
echo "Latest tailwind-cli-extra version: ${EXTRA_RELEASE_VERSION}"

echo "Installing tailwind-cli-extra: ${EXTRA_RELEASE_VERSION} - ${EXTRA_ASSET}"
EXTRA_URL="https://github.com/dobicinaitis/tailwind-cli-extra/releases/download/${EXTRA_RELEASE_VERSION}/${EXTRA_ASSET}"
echo "Downloading tailwindcss-extra from ${EXTRA_URL}"
curl -sLO "${EXTRA_URL}"
mv "${EXTRA_ASSET}" "$INSTALL_PATH/tailwindcss-extra"
chmod +x "$INSTALL_PATH/tailwindcss-extra"
$INSTALL_PATH/tailwindcss-extra --help
```

I even configured a specific step to show the installation results and what `trunk` sees as the available tools locally.

```sh
Run ./scripts/install-tailwindcss.sh
  ./scripts/install-tailwindcss.sh
  trunk tools show
  shell: sh -e {0}
  env:
    CARGO_TERM_COLOR: always
    SQLX_OFFLINE: true
    PATH: /usr/local/cargo/bin:
  
OS: linux
Architecture: x86_64
Fetching latest tailwind-cli release version...
Latest tailwind-cli version: v4.0.15
Installing tailwind-cli: v4.0.15 - tailwindcss-linux-x64
Downloading tailwindcss from https://github.com/tailwindlabs/tailwindcss/releases/download/v4.0.15/tailwindcss-linux-x64
≈ tailwindcss v4.0.15

Usage:
  tailwindcss [--input input.css] [--output output.css] [--watch] [options…]

Options:
  -i, --input ··········· Input file
  -o, --output ·········· Output file [default: `-`]
  -w, --watch ··········· Watch for changes and rebuild as needed
  -m, --minify ·········· Optimize and minify the output
      --optimize ········ Optimize the output without minifying
      --cwd ············· The current working directory [default: `.`]
  -h, --help ············ Display usage information
Fetching latest tailwind-cli-extra release version...
Latest tailwind-cli-extra version: v2.0.7
Installing tailwind-cli-extra: v2.0.7 - tailwindcss-extra-linux-x64
Downloading tailwindcss-extra from https://github.com/dobicinaitis/tailwind-cli-extra/releases/download/v2.0.7/tailwindcss-extra-linux-x64
≈ tailwindcss v4.0.15

Usage:
  tailwindcss [--input input.css] [--output output.css] [--watch] [options…]

Options:
  -i, --input ··········· Input file
  -o, --output ·········· Output file [default: `-`]
  -w, --watch ··········· Watch for changes and rebuild as needed
  -m, --minify ·········· Optimize and minify the output
      --optimize ········ Optimize the output without minifying
      --cwd ············· The current working directory [default: `.`]
  -h, --help ············ Display usage information
sass
    Installed Version: n/a
    Default Version: 1.69.5
    Download URL: https://github.com/sass/dart-sass/releases/download/1.69.5/dart-sass-1.69.5-linux-x64.tar.gz
    Location: n/a

tailwindcss
    Installed Version: n/a
    Default Version: 3.3.5
    Download URL: https://github.com/tailwindlabs/tailwindcss/releases/download/v3.3.5/tailwindcss-linux-x64
    Location: n/a

tailwindcss-extra
    Installed Version: n/a
    Default Version: 1.7.25
    Download URL: https://github.com/dobicinaitis/tailwind-cli-extra/releases/download/v1.7.25/tailwindcss-extra-linux-x64
    Location: n/a

wasm-bindgen
    Installed Version: 0.2.100
    Default Version: 0.2.89
    Download URL: https://github.com/rustwasm/wasm-bindgen/releases/download/0.2.89/wasm-bindgen-0.2.89-x86_64-unknown-linux-musl.tar.gz
    Location: /usr/local/cargo/bin/wasm-bindgen

wasm-opt
    Installed Version: n/a
    Default Version: version_116
    Download URL: https://github.com/WebAssembly/binaryen/releases/download/version_116/binaryen-version_116-x86_64-linux.tar.gz
    Location: n/a
```

You can see that `trunk` sees `wasm-bindgen` in the `/usr/local/cargo/bin/` path, the same location to where I've installed the tailwind CLIs, and the same path I ensured was in `PATH` for the `trunk` tasks. Yet, it still fails to find them.

This shouldn't be a big deal because it can download and install them on its own, right? Wrong! It fails to download and says the header was incorrect.-it didn't get the file it expected.

```sh
  trunk build --release
  shell: sh -e {0}
  env:
    CARGO_TERM_COLOR: always
    SQLX_OFFLINE: true
    CACHE_ON_FAILURE: false
    CARGO_INCREMENTAL: 0
    PATH: /usr/local/cargo/bin:
2025-03-23T20:03:07.034202Z  INFO Starting trunk 0.21.9
2025-03-23T20:03:07.034580Z  INFO starting build
   Compiling unicode-ident v1.0.18
   Compiling proc-macro2 v1.0.94
   Compiling wasm-bindgen-shared v0.2.100
   Compiling log v0.4.26
2025-03-23T20:03:08.745729Z  INFO downloading tailwindcss-extra version="2.0.7"
   Compiling rustversion v1.0.20
   Compiling bumpalo v3.17.0
   Compiling once_cell v1.21.1
   Compiling wasm-bindgen v0.2.100
   Compiling cfg-if v1.0.0
   Compiling serde v1.0.219
   Compiling memchr v2.7.4
2025-03-23T20:03:09.480512Z  INFO installing tailwindcss-extra
   Compiling quote v1.0.40
   Compiling itoa v1.0.15
   Compiling syn v2.0.100

... omitted for brevity ...

   Compiling shared v0.1.0 (/__w/RustyGPT/RustyGPT/shared)
   Compiling yewdux v0.11.0
   Compiling yew-router v0.18.0
   Compiling i18nrs v0.1.3
   Compiling yew-hooks v0.3.3
   Compiling yew_icons v0.8.0
   Compiling frontend v0.1.0 (/__w/RustyGPT/RustyGPT/frontend)
    Finished `release` profile [optimized] target(s) in 53.32s
2025-03-23T20:04:02.673301Z ERROR error
error from build pipeline

Caused by:
    0: HTML build pipeline failed (1 errors), showing first
    1: error from asset pipeline
    2: Could not extract files
    3: error while getting archive entry
    4: invalid gzip header
2025-03-23T20:04:02.673325Z ERROR error from build pipeline
2025-03-23T20:04:02.673334Z  INFO   1: HTML build pipeline failed (1 errors), showing first
2025-03-23T20:04:02.673336Z  INFO   2: error from asset pipeline
2025-03-23T20:04:02.673338Z  INFO   3: Could not extract files
2025-03-23T20:04:02.673339Z  INFO   4: error while getting archive entry
2025-03-23T20:04:02.673340Z  INFO   5: invalid gzip header
Error: Process completed with exit code 1.
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "trunk"
-version = "0.21.9"
+version = "0.21.10"
 edition = "2021"
 description = "Build, bundle & ship your Rust WASM application to the web."
 license = "MIT/Apache-2.0"
@@ -51,6 +51,7 @@ open = "5"
 oxipng = "9"
 parking_lot = "0.12"
 rand = "0.9.0"
+regex = "1"
 remove_dir_all = "1"
 reqwest = { version = "0.12", default-features = false, features = ["stream", "trust-dns"] }
 schemars = { version = "0.8", features = ["derive"] }

--- a/src/tools.rs
+++ b/src/tools.rs
@@ -183,7 +183,7 @@ impl Application {
 
     /// Format the output of version checking the app.
     pub(crate) fn format_version_output(&self, text: &str) -> Result<String> {
-        let regex_tailwind = Regex::new(r"(?m)^.+?v((?:\d+?\.?)+)$")
+        let regex_tailwind = Regex::new(r"(?m)^.+?v((?:\d+?\.?)+).*$")
             .context("failed to compile regex for tailwindcss version")?;
 
         let text = text.trim();
@@ -841,4 +841,75 @@ mod tests {
         "tailwindcss-extra v1.7.25",
         "1.7.25"
     );
+
+    // Additional format_version_output tests for TailwindCSS and TailwindCssExtra
+    table_test_format_version!(
+        tailwindcss_with_multiline_output,
+        Application::TailwindCss,
+        "Some other text\ntailwindcss v3.4.0",
+        "3.4.0"
+    );
+
+    table_test_format_version!(
+        tailwindcss_with_long_version,
+        Application::TailwindCss,
+        "tailwindcss v3.4.0.12345",
+        "3.4.0.12345"
+    );
+
+    table_test_format_version!(
+        tailwindcss_extra_with_multiline_output,
+        Application::TailwindCssExtra,
+        "≈ tailwindcss v4.0.14
+
+Usage:
+  tailwindcss [--input input.css] [--output output.css] [--watch] [options…]
+
+Options:
+  -i, --input ··········· Input file
+  -o, --output ·········· Output file [default: `-`]
+  -w, --watch ··········· Watch for changes and rebuild as needed
+  -m, --minify ·········· Optimize and minify the output
+      --optimize ········ Optimize the output without minifying
+      --cwd ············· The current working directory [default: `.`]
+  -h, --help ············ Display usage information",
+        "4.0.14"
+    );
+
+    table_test_format_version!(
+        tailwindcss_with_prefixed_text,
+        Application::TailwindCss,
+        "prefix-tailwindcss v3.3.3",
+        "3.3.3"
+    );
+
+    // Tests for error cases
+    #[test]
+    fn test_tailwindcss_missing_version() {
+        let app = Application::TailwindCss;
+        let result = app.format_version_output("tailwindcss without version");
+        assert!(result.is_err());
+    }
+
+    #[test]
+    fn test_tailwindcss_malformed_version() {
+        let app = Application::TailwindCss;
+        let result = app.format_version_output("tailwindcss vX.Y.Z");
+        assert!(result.is_err());
+    }
+
+    #[test]
+    fn test_tailwindcss_empty_version() {
+        let app = Application::TailwindCss;
+        let result = app.format_version_output("");
+        assert!(result.is_err());
+    }
+
+    #[test]
+    fn test_regex_captures_only_numbers() {
+        let app = Application::TailwindCss;
+        let result = app.format_version_output("tailwindcss v3.4.0-beta.1");
+        // Should only capture the numeric part, not the -beta.1
+        assert_eq!(result.unwrap_or_default(), "3.4.0");
+    }
 }

--- a/src/tools.rs
+++ b/src/tools.rs
@@ -7,6 +7,7 @@ use anyhow::{anyhow, bail, ensure, Context, Result};
 use directories::ProjectDirs;
 use futures_util::stream::StreamExt;
 use once_cell::sync::Lazy;
+use regex::Regex;
 use std::collections::HashMap;
 use std::path::PathBuf;
 use tokio::fs::File;
@@ -182,6 +183,9 @@ impl Application {
 
     /// Format the output of version checking the app.
     pub(crate) fn format_version_output(&self, text: &str) -> Result<String> {
+        let regex_tailwind = Regex::new(r"(?m)^.+?v((?:\d+?\.?)+)$")
+            .context("failed to compile regex for tailwindcss version")?;
+
         let text = text.trim();
         let formatted_version = match self {
             Application::Sass => text
@@ -189,18 +193,14 @@ impl Application {
                 .next()
                 .with_context(|| format!("missing or malformed version output: {}", text))?
                 .to_owned(),
-            Application::TailwindCss => text
-                .lines()
-                .find(|s| !str::is_empty(s))
-                .and_then(|s| s.split(" v").nth(1))
-                .with_context(|| format!("missing or malformed version output: {}", text))?
-                .to_owned(),
-            Application::TailwindCssExtra => text
-                .lines()
-                .find(|s| !str::is_empty(s))
-                .and_then(|s| s.split(" v").nth(1))
-                .with_context(|| format!("missing or malformed version output: {}", text))?
-                .to_owned(),
+            Application::TailwindCss | Application::TailwindCssExtra => {
+                let caps = regex_tailwind
+                    .captures(text)
+                    .with_context(|| format!("missing or malformed version output: {}", text))?;
+                caps.get(1)
+                    .map(|m| m.as_str().to_owned())
+                    .with_context(|| format!("missing capture group in version output: {}", text))?
+            }
             Application::WasmBindgen => text
                 .split(' ')
                 .nth(1)
@@ -448,16 +448,8 @@ async fn install(app: Application, archive_file: File, target_directory: PathBuf
 
     let target_directory_clone = target_directory.clone();
     tokio::task::spawn_blocking(move || {
-        let mut archive = if app == Application::Sass && cfg!(target_os = "windows") {
-            Archive::new_zip(archive_file)?
-        } else if app == Application::TailwindCss
-            || (app == Application::TailwindCssExtra
-                && (cfg!(target_os = "macos") || cfg!(target_os = "windows")))
-        {
-            Archive::new_none(archive_file)
-        } else {
-            Archive::new_tar_gz(archive_file)
-        };
+        let mut archive = Archive::new_from_header(archive_file)
+            .context("failed to determine archive type from header")?;
         archive.extract_file(app.path(), &target_directory)?;
 
         for path in app.extra_paths() {
@@ -548,7 +540,7 @@ async fn get_http_client(
 mod archive {
     use std::fmt::Display;
     use std::fs::{self, File};
-    use std::io::{self, BufReader, BufWriter, Read, Seek};
+    use std::io::{self, BufReader, BufWriter, Read, Seek, SeekFrom};
     use std::path::Path;
 
     use anyhow::{Context, Result};
@@ -575,6 +567,34 @@ mod archive {
 
         pub fn new_none(file: File) -> Self {
             Self::None(file)
+        }
+
+        /// Creates a new Archive by inspecting the file header.
+        /// - If the header matches a ZIP file ("PK"), returns `new_zip`.
+        /// - If the header matches a GZIP file (0x1F, 0x8B), returns `new_tar_gz`.
+        /// - Otherwise, returns `new_none`.
+        pub fn new_from_header(mut file: File) -> Result<Self> {
+            let mut header = [0u8; 4];
+            let bytes_read = {
+                let mut reader = BufReader::new(&mut file);
+                reader.read(&mut header).context("failed to read header")?
+            };
+
+            // Reset the file pointer to the beginning after reading the header.
+            file.seek(SeekFrom::Start(0))
+                .context("failed to rewind file")?;
+
+            if bytes_read >= 2 {
+                // ZIP header: 'P' (0x50) and 'K' (0x4B)
+                if header[0] == 0x50 && header[1] == 0x4B {
+                    return Self::new_zip(file);
+                }
+                // GZIP header: 0x1F and 0x8B (used for tar.gz)
+                if header[0] == 0x1F && header[1] == 0x8B {
+                    return Ok(Self::new_tar_gz(file));
+                }
+            }
+            Ok(Self::new_none(file))
         }
 
         pub fn extract_file(&mut self, file: &str, target_directory: &Path) -> Result<()> {

